qdr/_ranker.cc

```diff
@@ -0,0 +1,416 @@
+
+#include <math.h>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <stdexcept>
+
+
+typedef std::unordered_map<std::string, std::pair<uint64_t, uint64_t> >
+    counts_t;
+
+
+// count of word occurances in a doc
+typedef std::unordered_map<std::string, uint64_t> word_counts_t;
+
+// a tokenized document
+typedef std::vector<std::string> doc_t;
+
+word_counts_t count_words(const doc_t& document)
+{
+    /// Given a vector of tokens, return a map token -> count in vector
+    word_counts_t ret(document.size() * 3);
+    for (doc_t::const_iterator it = document.begin();
+        it != document.end(); ++it)
+    {
+        word_counts_t::iterator got = ret.find(*it);
+        if (got == ret.end())
+        {
+            // first time we've seen this word
+            ret[*it] = 1;
+        }
+        else
+            ret[*it] += 1;
+    }
+    return ret;
+}
+
+struct scores_t {
+    double tfidf;
+    double bm25;
+    double lm_jm;
+    double lm_dirichlet;
+    double lm_ad;
+};
+
+struct lm_scores_t {
+    double jm;
+    double dirichlet;
+    double ad;
+};
+
+class QDR
+{
+    public:
+        QDR(counts_t& counts, uint64_t total_docs);
+        ~QDR();
+
+        // compute the similarity scores
+        scores_t score(doc_t& document, doc_t& query);
+
+        // compute similiary scores for a single doc but list of queries
+        std::vector<scores_t> score_batch(doc_t& document,
+            std::vector<doc_t>& queries);
+
+        // get the IDF for a given word
+        double get_idf(const std::string& word);
+
+    private:
+        counts_t counts;
+        uint64_t total_docs;
+        uint64_t nwords_vocab;
+        uint64_t nwords;
+
+        // useful constant for language model
+        double n2p1;
+
+        // compute the TF-IDF similarity
+        double tfidf(const word_counts_t& doc_counts,
+            const word_counts_t& query_counts);
+
+        // BM25
+        double okapi_bm25(const word_counts_t& doc_counts,
+            const word_counts_t& query_counts,
+            const double& Ld);
+
+        // language model
+        lm_scores_t lm(const word_counts_t& doc_counts,
+            const word_counts_t& query_counts,
+            const double& sum_count_w_given_doc);
+
+        // low level function used by public methods
+        scores_t score_single(const word_counts_t& doc_counts,
+            const word_counts_t& query_counts,
+            const double& nwords_document);
+
+        // disable some default constructors
+        QDR();
+        QDR& operator= (const QDR& other);
+        QDR(const QDR& other);
+
+};
+
+QDR::QDR(counts_t& counts_in, uint64_t total_docs)
+{
+    this->total_docs = total_docs;
+
+    // copy the counts data to private member and add up the total
+    // number of words and words in vocab
+    nwords_vocab = 0;
+    nwords = 0;
+    counts.clear();
```

Comment:
This invocation of `clear` is superfluous. A newly constructed `unordered_map` is by definition empty.